### ISSUE TITLE
Is tensorflow consuming much more memory than torch?

### ISSUE THREAD

**Issue by yxchng** [tensorflow/tensorflow#4709]
I am trying to replicate the stacked hourglass architecture which is implemented in torch, https://github.com/anewell/pose-hg-train/. However, I am unable to train it using settings in the opts.lua file which set the train batch size as 6, number of train iteration as 8000, test batch size as 1 and number of test iteration as 1000. I understand that I can reduce the batch size or iteration number to fit the network. However, I just want to know if tensorflow is consuming more memory than torch, if the same architecture and settings can be implemented in tensorflow.

**Comment by yaroslavvb** [comment]
They have similar memory usage in general, although there are some special cases where it may differ substantially. You can see track down memory usage by looking at memory allocator statistics you get when running with FULL_TRACING options (see https://github.com/tensorflow/tensorflow/issues/1824#issuecomment-225754659
)

**Comment by yxchng** [comment]
do you what are the special cases?

**Comment by yaroslavvb** [comment]
calling `tf.gradients` on graph which has `tf.while` loop can take unexpectedly large amounts of memory

**Comment by asimshankar** [comment]
@yxchng : Were you able to obtain memory allocator statistics mentioned above?

**Comment by redst4r** [comment]
@yaroslavvb : can you elaborate on using FULL_TRACING to get memory allocator stats?
I managed to get the timing information via:
```python
from tensorflow.python.client import timeline
run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()
sess.run(some_stuff, options=run_options, run_metadata=run_metadata)

# for visualization
tl = timeline.Timeline(run_metadata.step_stats)
ctf = tl.generate_chrome_trace_format()
with open('timeline.json', 'w') as f:
    f.write(ctf)
```

But how would I assess the memory allocation from `run_metadata`?

**Comment by yaroslavvb** [comment]
If you print run_metadata, you'll see messages like this:

      output {
        tensor_description {
          dtype: DT_FLOAT
          shape {
          }
          allocation_description {
            requested_bytes: 4
            allocator_name: "cpu"
            ptr: 4493467328
          }
        }
      }

That means 4 bytes got allocated

**Comment by yaroslavvb** [comment]
You can also get raw memory allocator messages in your log like this:
http://stackoverflow.com/questions/36331419/tensorflow-how-to-measure-how-much-gpu-memory-each-tensor-takes/36505898#36505898

**Comment by redst4r** [comment]
thanks for the hint, I'll play around with that

**Comment by redst4r** [comment]
found this very convenient `show_memory` flag, which allows you to display mem-allocation over time in chrome-trace:

```python
ctf = tl.generate_chrome_trace_format(show_memory=True)
```

**Comment by aselle** [comment]
@yxchng have you come to any actionable bugs or issues with memory usage after looking at it more?

**Comment by yxchng** [comment]
@aselle I am busy with other stuff at the moment. I will try to look into it as soon as possible. Really sorry.

**Comment by aselle** [comment]
Automatically closing due to lack of recent activity, we will reopen if further information becomes available. Thanks!
